### ISSUE TITLE
How exactly to specify proteotypicity?

### ISSUE THREAD

**Issue by neely** [vdemichev/DiaNN#793]
This is more of a feature requests than an issue. The documentation for DIA-NN is great, but somehow I found this one aspect to be unclear: how do you specify strict proteotypicity for quant?

Example problem: The ABRF sPRG has a suite of samples with cow, human and trout proteomes mixed at different levels. Due to high degree of homology of some/many proteins, only proteotypic peptides should be used for quantification. Added wrinkle, the best fasta are not UniProt, and therefore we can't assume a "GN=" field in sequence header. (this isn't actually true in this case, but RefSeq fasta are my specific use-case)

The documentation says "Protein inference this setting primarily affects proteotypicity definition" but it doesn't explicitly say how it affects this or how to select strict proteotypicity. Utilizing the command line arguments of --relaxed-prot-inf --pg-level seems to be a work around to set protein inference, but again, I am not sure what setting would yield proteotypic. 

Further reading I thought maybe it would be in the unique_genes_matrix, or possibly something to be done to the matrix downstream of searching?

I apologize for being dense, but the recent sPRG study, as well as [a recent mixed human cow quant DIA-NN based study](https://www.biorxiv.org/content/10.1101/2023.08.28.555225v1) made me want specific instructions.

**Comment by vdemichev** [comment]
Hi Ben,

For UniProt, DIA-NN has the Genes (species-specific) protein inference mode, for the case when need to distinguish orthologues.

Genes.MaxLFQ.Unique would be the gene-product quant using proteotypic peptides only.
Or indeed the unique_genes matrix, which contains the same quantities.

Yes, for non-UniProt, I guess the solution will be to use some R or Python package to reformat the FASTAs first.

If you'd like to have not gene-level but sequence id-level proteotypicity, and only use thus defined proteotypic peptides for quant, the easiest will be to just reformat the FASTA (in any format) to add GN=[sequence ID for the protein] to trick DIA-NN into thinking that sequence IDs are also gene IDs. 

Would the above address the issue?

Best,
Vadim

**Comment by vdemichev** [comment]
Btw, just filtering the main report and quantifying proteins in R using peptides of choice is also always an option.

**Comment by neely** [comment]
Vadim,

Thanks for a very quick response. 

Adding GN to refseq isn't the worst idea (and I actually have a tool doing this). The problem, even with pretty good non-model annotations, is I still run into LOC#####, even though it might be a real name. In this case it wouldn't know they were the same gene.

I didn't realize the Genes (species-specific) protein inference mode distinguished orthologues. So if I had a horse and human database and say was looking at GN=HP, which might have shared peptide IDs, then the Genes.MaxLFQ.Unique column would be the gene-product quant of horse and human HP separately using proteotypic peptides only. So in this bad example, it would know one HP was one entry (horse) and one HP was another entry (human), each getting their own row?

**Comment by vdemichev** [comment]
Yes, Genes (species-specific) just adds the species suffix to the gene name. So instead of HP you get HP_HUMAN reported in the Genes column. And respectively Genes.MaxLfQ.Unique calculated for it. The species needs to be encoded in the fasta header same as in uniprot format

**Comment by neely** [comment]
Okay, and not to beleaguer the point, but is it using the alt accession name (which appends the 5 letter genus species abbreviation) or adding a species name to the “GN=“ field? Haptoglobin is a good example. Its alt accession name is HPT_HUMAN but it’s GN=HP and OS=Homo sapiens. This clarification will help me know what I need to add to the refseq to make it work.

**Comment by vdemichev** [comment]
When in Genes (species-specific) mode, DIA-NN reads a UniProt header, like
">sp|P02768|ALBU_HUMAN Albumin OS=Homo sapiens OX=9606 GN=ALB PE=1 SV=2"

and extracts '_HUMAN' from ALBU_HUMAN as the species name, and then concatenates it with whatever is after GN=, i.e. the header is read if it was 
">sp|P02768|ALBU_HUMAN Albumin OS=Homo sapiens OX=9606 GN=ALB_HUMAN PE=1 SV=2"

**Comment by neely** [comment]
Thanks for clarifying it is using the 5 letter species code from the alt accession, not the OS field. So that’s a bit more involved than just adding a GN and OS field. Still, I can make a full fledged refseq to uniprot converter. Before I set off on this, can you confirm your parsing is using the pipes of the fasta header: so after second pipe up to the space has the alt accession which ends in “_5 letter code”, and your parsing will allow for more than two letters before the first pipe (I want to use “>refseq|…”)?

**Comment by vdemichev** [comment]
OK, I did not fully understand the question :) 

This is the full parsing code:

```
void annotation_sgd(std::string& header, std::string& id, std::string& name, std::string& gene, std::string& description, bool* swissprot) {
	*swissprot = (header.find("Verified ORF") != std::string::npos);
	int start = 0, end2;
	auto end = header.find_first_of(' ');
	if (end == std::string::npos) id = fast_trim(header.substr(start + 1)), name = description = "";
	else {
		id = fast_trim(header.substr(start + 1, end - start - 1));
		for (end2 = end + 1; end2 < header.size() && header[end2] != ' '; end2++);
		end2 = header.find_first_of(' ', end2);
		name = fast_trim(header.substr(end + 1, end2 - end - 1));
		end = header.find_first_of('"', end2);
		if (end == std::string::npos) description = "";
		else {
			end2 = header.find_first_of('"', end + 1);
			if (end2 == std::string::npos) description = "";
			else description = header.substr(end + 1, end2 - end - 1);
		}
	}
	gene = name;
}

void annotation(std::string& header, std::string& id, std::string& name, std::string& gene, std::string& description, bool* swissprot) {
	if (header.size() < 3) {
		id = "", name = "", gene = "", description = "", * swissprot = false;
		return;
	}
	auto start = header.find("|");
	if (start == std::string::npos || (start != 3 && !(header[1] == 's' && header[2] == 'p') && !(header[1] == 't' && header[2] == 'r'))) {
		annotation_sgd(header, id, name, gene, description, swissprot);
		return;
	}
	auto end = header.find("|", start + 1);
	if (end == std::string::npos) end = header.find(" ", start + 1), name = description = "";
	else {
		auto end2 = header.find(" ", end + 1);
		if (end2 != std::string::npos) {
			name = fast_trim(header.substr(end + 1, end2 - end - 1));
			auto end3 = header.find("=", end2 + 1);
			if (end3 != std::string::npos) {
				while (header[end3] != ' ' && end3 > end2 + 2) end3--;
				if (end3 > end2 + 2) description = fast_trim(header.substr(end2 + 1, end3 - end2 - 1));
				else description = "";
			} else description = fast_trim(header.substr(end2 + 1));
		} else name = fast_trim(header.substr(end + 1)), description = "";
	}
	if (end != std::string::npos) id = fast_trim(header.substr(start + 1, end - start - 1));
	else id = fast_trim(header.substr(start + 1));
	start = header.find("GN=");
	if (start == std::string::npos) gene = "";
	else {
		end = header.find(" ", start + 3);
		if (end != std::string::npos) gene = fast_trim(header.substr(start + 3, end - start - 3));
		else gene = fast_trim(header.substr(start + 3));
	}
	*swissprot = (header[1] == 's' && header[2] == 'p');
	if (SpeciesGenes && gene.size() && name.size()) {
		int pos = name.find_last_of('_');
		if (pos != std::string::npos) gene += name.substr(pos);
	}
}
```

**Comment by vdemichev** [comment]
std::string::npos means not found in the string, SpeciesGenes is Genes (species-specific) mode.

**Comment by vdemichev** [comment]
Not exactly:
For uniprot, it's like this:
>sp|sequence-id-here|name_SUFFIX ... GN=gene
Where it can be 'tr' instead of 'sp' and ... can be anything.

If it's not sp and not tr, unless it is formatted SGD (yeast)-style, it will only extract the sequence ID and nothing else.

**Comment by vdemichev** [comment]
If you want to encode some protein-origin info, instead of 'refseq' (will not be interpreted by DIA-NN) can concatenate this info to the sequence id

**Comment by neely** [comment]
Thanks for clarifying. So I need to completely spoof the uniprot header, and I’ll even use ole sp. Then I can perform proper quant on a mixed species sample by using only proteotypic peptides (Gene level inference and use Genes.MaxLfQ.Unique column) for quant?

Aside: there’s so many species that have great genome annotations on NCBI but aren’t in UniProt (or those there are outdated or poorly imported), yet many proteomic tools only fully function with uniprot fasta (or xml).

I do appreciate the help on clarifying proteotypic settings and sorry to conflate this with an issue of sequence header parsing.

**Comment by vdemichev** [comment]
I will add an option in DIA-NN to fill the 'Genes' column with sequence IDs. Since the latter are read correctly from any FASTA, this should work nicely.

**Comment by neely** [comment]
Again, much appreciated and thank for digressing to header issues. I’ll still need to add “sp|” before and “|” after the current NCBI sequence ID right? That’s super easy. [Here](https://github.com/neely/ChatGPT/tree/main/addGeneSymbol2RefSeq) is the chatgpt-aided python script solution for me adding “GN=“. I really can easily add to this enough to make it work at least with DIA-NN like we’re discussing.
